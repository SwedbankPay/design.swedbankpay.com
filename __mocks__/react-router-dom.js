import React from "react";

const rrd = require("react-router-dom"); // Just render plain div with its children

// eslint-disable-next-line react/display-name
rrd.BrowserRouter = ({ children }) => <div>{children}</div>;

module.exports = rrd;
