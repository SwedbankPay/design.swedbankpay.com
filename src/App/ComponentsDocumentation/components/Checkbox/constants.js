import React from "react";
import Checkbox from "@components/FormComponents/Checkbox";

const options = [
    {
        id: "radio-example-1",
        label: "Label"
    },
    {
        id: "radio-example-2",
        label: "Label"
    }
];

export const checkboxShowCase = {
    id: "no-tabs",
    elements: [
        {
            component: <Checkbox groupTitle="Group label" name="checkbox-group-example" options={options} group hintExpander={true} />,
            options: {
                checkbox: [
                    {
                        title: "Label add on",
                        inputs: [
                            {
                                id: "checkbox_labels",
                                name: "Is optional",
                                value: {
                                    optional: true
                                }
                            },
                            {
                                id: "hint_text",
                                name: "Hint text",
                                value: {
                                    hintText: "Hint text",
                                    hintTextId: "hintTextCheckboxExample"
                                }
                            },
                            {
                                id: "hint_expander",
                                name: "Hint expander",
                                value: {
                                    expandingHintTitle: "Expanding hint text",
                                    expanderId: "expander-showcase"
                                }
                            }
                        ]
                    }
                ],
                radio: [
                    {
                        id: "state_radio",
                        title: "State modifiers",
                        values: [
                            {
                                name: "Enabled",
                                value: {
                                    disabled: false
                                }
                            },
                            {
                                name: "Disabled",
                                value: {
                                    disabled: true
                                }
                            },
                            {
                                name: "Error",
                                value: {
                                    errorMessage: "Descriptive helpful error message."
                                }
                            }
                        ]
                    }
                ]
            },
            title: "Checkbox group",
            description: "The default setup of the checkbox group has a label on top which describes the checkbox group and each individual checkbox is then followed by a label which clearly describes the option and what checking the box represents."
        }
    ]
};

export const guidelineOptions = [
    {
        id: "radio-guidelines-example-1",
        label: "Label"
    },
    {
        id: "radio-guidelines-example-2",
        label: "Label"
    },
    {
        id: "radio-guidelines-example-3",
        label: "Label"
    }
];

export const developerDocOptions = [
    {
        id: "error-example-1",
        label: "Label"
    },
    {
        id: "error-example-2",
        label: "Label"
    },
    {
        id: "error-example-3",
        label: "Label"
    }
];

export const doOptions = [
    {
        id: "radio-do-example-1",
        label: "Small"
    },
    {
        id: "radio-do-example-2",
        label: "Medium"
    },
    {
        id: "radio-do-example-3",
        label: "Large"
    }
];

export const dontOptions = [
    {
        id: "radio-dont-example-1",
        label: "Super duper tiny small"
    },
    {
        id: "radio-dont-example-2",
        label: "Large"
    },
    {
        id: "radio-dont-example-3",
        label: "Small"
    }
];
