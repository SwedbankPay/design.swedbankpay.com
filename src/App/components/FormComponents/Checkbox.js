import React from "react";
import PropTypes from "prop-types";

const Checkbox = ({ id, checked, name, disabled, groupTitle, label, errorMessage, helpBlock, optional, required, group, options, className }) => {
    const attrs = {
        type: "checkbox",
        id: id || null,
        name,
        disabled: disabled || null,
        defaultChecked: checked || null,
        required,
        "aria-describedby": helpBlock || null
    };

    return (
        <>
            {group ? <form>
                <fieldset className={`checkbox-group${errorMessage ? " has-error" : ""}`} disabled={disabled}>
                    <legend>
                        {groupTitle}
                        {optional && <>{"\n"}
                            <span>(optional)</span>
                        </>}
                    </legend>
                    {options.map(({ label, id }, i) => <div className="checkbox" key={i}>{"\n"}
                        <input {...attrs} id={id} />{"\n"}
                        <label htmlFor={id}>{label}</label>{"\n"}
                    </div>)}
                    {errorMessage && <><div className="help-block"><i className="material-icons">warning</i>{errorMessage}</div>{"\n"}</>}
                    {helpBlock && <><p id="hint-text" className="hint-text">{helpBlock}</p>{"\n"}</>}
                </fieldset>
            </form>
                : <>
                    <div className={`checkbox${className ? ` ${className}` : ""}${errorMessage ? " has-error" : ""}`}>{"\n"}
                        <input {...attrs} />{"\n"}
                        <label htmlFor={id} className="d-block">{label}</label>{"\n"}
                    </div>
                </>
            }
        </>
    );
};

Checkbox.propTypes = {
    id: PropTypes.string,
    checked: PropTypes.bool,
    disabled: PropTypes.bool,
    label: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.object
    ]),
    name: PropTypes.string,
    errorMessage: PropTypes.string,
    require: PropTypes.bool,
    group: PropTypes.bool,
    groupTitle: PropTypes.string,
    options: PropTypes.array,
    optional: PropTypes.bool,
    helpBlock: PropTypes.string,
};

export default Checkbox;
