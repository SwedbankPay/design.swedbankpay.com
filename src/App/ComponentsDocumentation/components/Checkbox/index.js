import React from "react";
import { Link } from "react-router-dom";

import { ComponentPreview, DocContainer } from "@docutils";
import CheckboxComponent from "@components/FormComponents/Checkbox";
import CodeTags from "@components/CodeTags";
import { checkboxShowCase, guidelineOptions, developerDocOptions, doOptions, dontOptions } from "./constants";

const HowTo = () => (
    <section>
        <h3>How to setup checkbox groups</h3>
        <div className="row">
            <div className="col-6">
                <div className="slab slab-plain slab-border-success h-100">
                    <span className="h3">Do</span>
                    <div className="d-flex flex-column align-items-center">
                        <div className="pt-3 pb-5">
                            <CheckboxComponent label="Button size" name="radio-guidelines" options={doOptions} group />
                        </div>
                        <p>Short and concise, similar length on the text is always preferred when setting up a checkbox group. Try to sort the options in a logical order.</p>
                    </div>
                </div>
            </div>
            <div className="col-6">
                <div className="slab slab-plain slab-border-error">
                    <span className="h3">Don&#x27;t</span>
                    <div className="d-flex flex-column align-items-center">
                        <div className="pt-3 pb-5 w-50">
                            <CheckboxComponent label="Button size" name="radio-guidelines" options={dontOptions} group />
                        </div>
                        <p>Avoid having labels longer than the rest when possible and displaying options without order.</p>
                    </div>
                </div>
            </div>
        </div>
    </section>
);

const Overview = () => (
    <section>
        <h2 id="overview">Overview</h2>

        <ComponentPreview language="html" showCasePanel showCasePanelAdvanced={checkboxShowCase} />

        <h3>When to consider something else</h3>
        <ul className="list list-bullet">
            <li>
            When you want to present multiple choices where the user can only select one specific choice, the <Link to="/components/radiobutton">radio button</Link> component should be used instead of checkboxes.
            </li>
        </ul>

        <HowTo />
    </section>
);

const ContentGuidelines = () => (
    <section>
        <h2 id="content-guidelines">Content guidelines</h2>
        <div className="slab slab-plain">
            <div className="d-flex flex-column align-items-center py-5">
                <CheckboxComponent label="Group heading" name="radio-guidelines" options={guidelineOptions} group />
            </div>
        </div>
        <ul className="list list-bullet">
            <li>The <b>Group heading</b> for the radio button group should be descriptive, explaining the context and what actions the user is choosing from. </li>
            <li>The <b>radio button label</b> should be short, to the point and clearly explain the option. It is also preferred to have fewer than three words per label as longer words could extend a label to multiple rows which could disrupt the reading experience.</li>
        </ul>
    </section>
);

const DeveloperDocumentation = () => (
    <section>
        <h2 id="developer-documentation">Developer documentation</h2>
        <span className="h3 d-inline-block mt-3">Disabled state</span>
        <p>Disable a checkbox by adding <CodeTags type="primary" code={"disabled"} /> attribute to the desired Checkbox. You can also disable the <CodeTags type="secondary" code={".form-group"} /> or <CodeTags type="primary" code={"fieldset"} /> in which the checkbox is located, refer to <Link to="/components/forms">forms</Link> for more info.</p>

        <ComponentPreview language="html" showCasePanel codeFigure removeOuterTag>
            <div className="d-flex justify-content-center">
                <CheckboxComponent id="disabled-example" label="Checkbox label" disabled/>
            </div>
        </ComponentPreview>

        <span className="h3 d-inline-block mt-3">Error state</span>
        <p>Insert the error message in the attribute <CodeTags type="primary" code={"data-error"} /> in the <CodeTags type="secondary" code={".help-block"} /> element. </p>
        <ComponentPreview language="html" showCasePanel codeFigure removeOuterTag>
            <div className="w-50 m-auto pl-5">
                <CheckboxComponent label="Group heading" name="checkbox-group-example" options={developerDocOptions} group errorMessage="Descriptive helpful error message." />
            </div>
        </ComponentPreview>
    </section>
);

const Checkbox = () => (
    <DocContainer>
        <p className="lead">The checkbox component should be used when the user are presented with a list of options where one or multiple options can be selected by the user.</p>
        <Overview />
        <ContentGuidelines />
        <DeveloperDocumentation />
    </DocContainer>
);

export default Checkbox;

/* for testing */
export { Overview, ContentGuidelines, HowTo };
