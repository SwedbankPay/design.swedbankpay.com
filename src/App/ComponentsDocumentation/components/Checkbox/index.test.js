import React from "react";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";

import Checkbox, { Overview, ContentGuidelines, DeveloperDocumentation, HowTo } from "./index";

describe("Component: Checkbox", () => {
    it("is defined", () => {
        expect(Checkbox).toBeDefined();
    });

    it.skip("renders", () => {
        const wrapper = renderer.create(<BrowserRouter><Checkbox /></BrowserRouter>);

        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    describe("Overview", () => {
        it("is defined", () => {
            expect(Overview).toBeDefined();
        });

        it.skip("renders", () => {
            const wrapper = renderer.create(<Overview />);

            expect(wrapper.toJSON()).toMatchSnapshot();
        });
    });

    describe("ContentGuidelines", () => {
        it("is defined", () => {
            expect(ContentGuidelines).toBeDefined();
        });

        it("renders", () => {
            const wrapper = renderer.create(<ContentGuidelines />);

            expect(wrapper.toJSON()).toMatchSnapshot();
        });
    });

    describe("DeveloperDocumentation", () => {
        it("is defined", () => {
            expect(DeveloperDocumentation).toBeDefined();
        });

        it("renders", () => {
            const wrapper = renderer.create(<DeveloperDocumentation />);

            expect(wrapper.toJSON()).toMatchSnapshot();
        });
    });

    describe("HowTo", () => {
        it("is defined", () => {
            expect(HowTo).toBeDefined();
        });

        it("renders", () => {
            const wrapper = renderer.create(<HowTo />);

            expect(wrapper.toJSON()).toMatchSnapshot();
        });
    });
});
