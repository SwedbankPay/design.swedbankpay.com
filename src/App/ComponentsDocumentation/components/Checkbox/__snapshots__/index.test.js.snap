// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Checkbox ContentGuidelines renders 1`] = `
<section>
  <h2
    id="content-guidelines"
  >
    Content guidelines
  </h2>
  <div
    className="slab slab-plain"
  >
    <div
      className="d-flex flex-column align-items-center py-5"
    >
      <Checkbox
        group={true}
        groupTitle="Group heading"
        name="radio-guidelines"
        options={
          Array [
            Object {
              "id": "radio-guidelines-example-1",
              "label": "Checkbox label",
            },
            Object {
              "id": "radio-guidelines-example-2",
              "label": "Checkbox label",
            },
            Object {
              "id": "radio-guidelines-example-3",
              "label": "Checkbox label",
            },
          ]
        }
      />
    </div>
  </div>
  <ul
    className="list list-bullet"
  >
    <li>
      The 
      <b>
        Group heading
      </b>
       should descriptive, explaining the context of the options to choose from. 
    </li>
    <li>
      The 
      <b>
        checkbox labels
      </b>
       should be short, to the point and clearly explain the option. It is also preferred to have fewer than three words per label as longer words could extend a label to multiple rows which could disrupt the reading experience.
    </li>
  </ul>
</section>
`;

exports[`Component: Checkbox DeveloperDocumentation renders 1`] = `
<section>
  <h2
    id="developer-documentation"
  >
    Developer documentation
  </h2>
  <span
    className="h3 d-inline-block mt-3"
  >
    Disabled state
  </span>
  <p>
    Disable a checkbox by adding 
    <Tag
      code={true}
      text="disabled"
      type="primary"
    />
     attribute to the desired Checkbox.
  </p>
  <ComponentPreview
    codeFigure={true}
    language="html"
    removeOuterTag={true}
    showCasePanel={true}
  >
    <div
      className="d-flex justify-content-center"
    >
      <Checkbox
        disabled={true}
        id="disabled-example"
        label="Checkbox label"
      />
    </div>
  </ComponentPreview>
  <span
    className="h3 d-inline-block mt-3"
  >
    Error state
  </span>
  <p>
    Insert the error message in the attribute 
    <Tag
      code={true}
      text="data-error"
      type="primary"
    />
     in the 
    <Tag
      code={true}
      text=".help-block"
      type="secondary"
    />
     element. 
  </p>
  <ComponentPreview
    codeFigure={true}
    language="html"
    removeOuterTag={true}
    showCasePanel={true}
  >
    <div
      className="m-auto pl-5 col-sm-12 error-state-container"
    >
      <Checkbox
        errorMessage="Descriptive helpful error message."
        group={true}
        groupTitle="Group heading"
        name="checkbox-group-example"
        options={
          Array [
            Object {
              "id": "error-example-1",
              "label": "Checkbox label",
            },
            Object {
              "id": "error-example-2",
              "label": "Checkbox label",
            },
            Object {
              "id": "error-example-3",
              "label": "Checkbox label",
            },
          ]
        }
      />
    </div>
  </ComponentPreview>
</section>
`;

exports[`Component: Checkbox HowTo renders 1`] = `
<section>
  <h3>
    How to setup checkbox groups
  </h3>
  <div
    className="row"
  >
    <div
      className="col-6"
    >
      <div
        className="slab slab-plain slab-border-success"
      >
        <span
          className="h3"
        >
          Do
        </span>
        <div
          className="d-flex flex-column align-items-center"
        >
          <div
            className="pt-3 pb-4"
          >
            <Checkbox
              group={true}
              groupTitle="Button size"
              name="radio-guidelines"
              options={
                Array [
                  Object {
                    "id": "radio-do-example-1",
                    "label": "Small",
                  },
                  Object {
                    "id": "radio-do-example-2",
                    "label": "Medium",
                  },
                  Object {
                    "id": "radio-do-example-3",
                    "label": "Large",
                  },
                ]
              }
            />
          </div>
        </div>
      </div>
      <p>
        Do use short labels of similar length and sort the options either in a logical or alphabetical order.
      </p>
    </div>
    <div
      className="col-6"
    >
      <div
        className="slab slab-plain slab-border-error"
      >
        <span
          className="h3"
        >
          Don't
        </span>
        <div
          className="d-flex flex-column align-items-center"
        >
          <div
            className="pt-3 pb-3 w-50"
          >
            <Checkbox
              group={true}
              groupTitle="Button size"
              name="radio-guidelines"
              options={
                Array [
                  Object {
                    "id": "radio-dont-example-1",
                    "label": "Super duper tiny small",
                  },
                  Object {
                    "id": "radio-dont-example-2",
                    "label": "Large",
                  },
                  Object {
                    "id": "radio-dont-example-3",
                    "label": "Small",
                  },
                ]
              }
            />
          </div>
        </div>
      </div>
      <p>
        Avoid long labels sorted in an illogical order.
      </p>
    </div>
  </div>
</section>
`;

exports[`Component: Checkbox Overview renders 1`] = `
<section>
  <h2
    id="overview"
  >
    Overview
  </h2>
  <ComponentPreview
    language="html"
    showCasePanel={true}
    showCasePanelAdvanced={
      Object {
        "elements": Array [
          Object {
            "component": <Checkbox
              group={true}
              groupTitle="Group heading"
              name="checkbox-group-example"
              options={
                Array [
                  Object {
                    "id": "radio-example-1",
                    "label": "Checkbox label",
                  },
                  Object {
                    "id": "radio-example-2",
                    "label": "Checkbox label",
                  },
                  Object {
                    "id": "radio-example-3",
                    "label": "Checkbox label",
                  },
                ]
              }
            />,
            "description": "The default setup of the checkbox group always has a label on top which describes the checkbox group and each individual checkbox is then followed by a label which clearly describes what the options and what checking the box represents",
            "options": Object {
              "checkbox": Array [
                Object {
                  "inputs": Array [
                    Object {
                      "id": "checkbox_labels",
                      "name": "Optional",
                      "value": Object {
                        "optional": true,
                      },
                    },
                    Object {
                      "id": "with_help_icon",
                      "name": "With help icon",
                      "value": Object {
                        "tooltip": true,
                      },
                    },
                  ],
                  "title": "Label add on",
                },
              ],
              "radio": Array [
                Object {
                  "id": "state_radio",
                  "title": "State modifiers",
                  "values": Array [
                    Object {
                      "name": "Enabled",
                      "value": Object {
                        "disabled": false,
                      },
                    },
                    Object {
                      "name": "Disabled",
                      "value": Object {
                        "disabled": true,
                      },
                    },
                    Object {
                      "name": "Error",
                      "value": Object {
                        "errorMessage": "Descriptive helpful error message.",
                      },
                    },
                  ],
                },
              ],
            },
            "title": "Checkbox group",
          },
        ],
        "id": "no-tabs",
      }
    }
  />
  <h3>
    When to consider something else
  </h3>
  <ul
    className="list list-bullet"
  >
    <li>
      When you want to present multiple choices where the user can only select one specific option, the 
      <Link
        to="/components/radio-button"
      >
        radio button
      </Link>
       component should be used instead of checkboxes.
    </li>
  </ul>
  <HowTo />
</section>
`;

exports[`Component: Checkbox renders 1`] = `
<DocContainer>
  <section
    id="checkbox-doc"
  >
    <p
      className="lead"
    >
      The checkbox component should be used when the user are presented with a list of options where one or multiple options can be selected by the user.
    </p>
    <Overview />
    <ContentGuidelines />
    <DeveloperDocumentation />
  </section>
</DocContainer>
`;
